File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.js
var subarraySum = function (nums, k) {
  // let sum = 0;
  // let count = 0;
  // let sumCountMap = new Map();
  // sumCountMap.set(0, 1);
  // for (let i = 0; i < nums.length; i++) {
  //   sum += nums[i];
  //   if (sumCountMap.has(sum - k)) {
  //     count += sumCountMap.get(sum - k);
  //   }
  //   if(sumCountMap.has(sum)){
  //     sumCountMap.set(sum,sumCountMap.get(sum)+1);
  //   }
  //   else{
  //     sumCountMap.set(sum,1);
  //   }
  // }
  // return count;
    let count = 0;
    let sum = 0; 
    let sumCountMap = new Map();
    sumCountMap.set(0,1);
    for(let i = 0 ; i < nums.length; i++){
        sum += nums[i];
        if(sumCountMap.has(sum-k)){
            count += sumCountMap.get(sum-k);
        }
        if(sumCountMap.has(sum)){
            sumCountMap.set(sum,sumCountMap.get(sum)+1)
        }
        else{
            sumCountMap.set(sum,1)
        }
    }
    return count;
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        sum_map = defaultdict(int)
        cum_sum = 0
        sum_map[0] = 1
        count = 0
        for i in range(len(nums)):
            cum_sum += nums[i]
            if cum_sum - k in sum_map:
                count += sum_map[cum_sum - k]

            sum_map[cum_sum] += 1

        return count
        